ospfd: add support for Topology Independent LFA (TI-LFA)

GalaxyGorilla opened the pull request:
This PR strives to implement TI-LFA for OSPFv2 and hence this is the OSPF counterpart to #7011. 

Experimental support for Topology Independent LFA (Loop-Free Alternate), see
for example 'draft-bashandy-rtgwg-segment-routing-ti-lfa-05'. The related
files are `ospf_ti_lfa.c/h`.

The current implementation is rather naive and does not support the advanced
optimizations suggested in e.g. RFC7490 or RFC8102. It focuses on providing
the essential infrastructure which can also later be used to enhance the
algorithmic aspects.

Supported features:

* Link and node protection
* Intra-area support
* Proper use of Prefix- and Adjacency-SIDs in label stacks
* Asymmetric weights (using reverse SPF)
* Non-adjacent P/Q spaces
* Protection of Prefix-SIDs

If configured for every SPF run the routing table is enriched with additional
backup paths for every prefix. The corresponding Prefix-SIDs are updated with
backup paths too within the OSPF SR update task.

Testing:

* unit test framework + tests for fancy topologies (focus on label stacks)
* topotests (focus on getting the label stacks as backup path into the RIB)

Signed-off-by: GalaxyGorilla <sascha@netdef.org>

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

riw777 commented on ospfd/ospf_ti_lfa.h:
```diff
@@ -0,0 +1,35 @@
+ * Copyright (C) 2020  NetDEF, Inc.
+ *                     Sascha Kattelmann
+ *
+ * This file is part of GNU Zebra.
```
should this be FR Routing?

GalaxyGorilla replied:
fixed (done similar to BFD code)

riw777 reviewed (commented):
Preliminary run at this ... So far looks good. Probably needs a lot more in depth look when it's closer to done, though. :-)

GalaxyGorilla commented on ospfd/ospf_ti_lfa.c:
```diff
@@ -0,0 +1,552 @@
+ *
+ * This file is part of GNU Zebra.
+ *
+ * GNU Zebra is free software; you can redistribute it and/or modify it
```
fixed (done similar to BDF code)

GalaxyGorilla commented:
> Preliminary run at this ... So far looks good. Probably needs a lot more in depth look when it's closer to done, though. :-)

Thanks @riw777 :)
just fyi: I'm currently working on adjacency SID support (somewhat easy but there's some support missing in the SR code) and also researching inter-area support.

pushpasis commented:
@donaldsharp You can add me to the reviewers list as well.

GalaxyGorilla commented:
CI:rerun

riw777 commented:
@GalaxyGorilla -- let me know when you're ready for review on this ... :-)

GalaxyGorilla commented:
> @GalaxyGorilla -- let me know when you're ready for review on this ... :-)

When the CI is green :)), the new commit fixes some discoveries by @rwestphal and I think I addressed all of them.

odd22 reviewed (commented):
Thanks for this awesome work. This PR looks good to me. 

I have only 2 comments about extra file and function

GalaxyGorilla commented on ospfd/ospf_sr.c:
```diff
@@ -159,6 +159,16 @@ static struct sr_node *sr_node_new(struct in_addr *rid)
 	return new;
 }
 
+/* Supposed to be used for testing */
+struct sr_node *ospf_sr_node_create(struct in_addr *rid)
+{
+	struct sr_node *srn;
+
+	srn = hash_get(OspfSR.neighbors, (void *)rid, (void *)sr_node_new);
+
+	return srn;
+}
+
```
🤔 that looks like some leftover, I'll try to get rid of it.

GalaxyGorilla replied:
Ok I checked the code. The problem here was that `OspfSR ` is defined statically in that file. So I don't see a way around this solution. If you have an idea please drop it here, I'll resolve the issue for now.

GalaxyGorilla commented on tests/ospfd/.gitignore:
```diff
@@ -0,0 +1,3 @@
+/*_afl/*
+test_ospf_spf
+core
```
Yea, the .gitignore needs to be aligned here AFAIK. Those are executables for the new unit tests.

GalaxyGorilla commented:
Rebased and I added another commit (95ce786) which is about P2MP handling. What I did there is quite hacky but currently I don't see a way around this. The problem is that we need to distinguish between P2P and P2MP without using interface data (because we don't have interface data from other nodes) and this turned out to be tricky just by using LSA data.

GalaxyGorilla commented:
Let's wait here until #7687 is merged, I wanna be sure that I did not break P2MP again ...

GalaxyGorilla commented:
#7687 is merged, CI is still green. @odd22 can you have a final look?

riw777 reviewed (approved):
looks good ... will wait to see if @odd22 takes a look, and to clear the conflicts before pushing

GalaxyGorilla commented:
conflicts resolved ...
